### ISSUE TITLE
[Question] Difference between a BackSolve Adjoint and a just a lax.scan

### ISSUE THREAD

**Issue by ASKabalan** [patrick-kidger/diffrax#572]
Hello,

I was wondering if `BackSolveAdjoint` is needed in case I have a solver that takes a ConstantStepSize

This is a very minimal example

```python
import jax
import jax.numpy as jnp
from jax import lax
import equinox.internal as eqxi
from functools import partial  

def euler_step(f, y, t, dt):

    return y + dt * f(y, t)

def integrate_ode_euler(f, y0, t0, t1, num_steps , kind):

    dt = (t1 - t0) / num_steps

    def step(y, t):
        """Single step for lax.scan."""
        y_next = euler_step(f, y, t, dt)
        return y_next, None
    results, _ = eqxi.scan(step, y0, jnp.arange(num_steps) , kind=kind )
    return results[-1]

# Example usage
def ode_func(y, t):
    """Example ODE: dy/dt = -y."""
    return -y

# Parameters
y0 = jnp.arange(1024.)  # Initial state (multi-dimensional)
t0 = 0.0                   # Initial time
t1 = 5.0                   # Final time
num_steps = 100            # Number of steps

@partial(jax.jit , static_argnums=(1))
def fn_DTO(y0 , num_steps):
    ys = integrate_ode_euler(ode_func, y0, t0, t1, num_steps , kind="checkpointed")
    return ys.sum()

@partial(jax.jit , static_argnums=(1))
def fn_OTD(y0 , num_steps):
    ys = integrate_ode_euler(ode_func, y0, t0, t1, num_steps , kind="lax")
    return ys.sum()


g_fn_DTO = jax.jit(jax.grad(fn_DTO , argnums=0) , static_argnums=1)
g_fn_OTD = jax.jit(jax.grad(fn_OTD , argnums=0) , static_argnums=1)

assert jnp.allclose(fn_DTO(y0 , num_steps), fn_OTD(y0 , num_steps))
assert jnp.allclose(g_fn_DTO(y0 , num_steps), g_fn_OTD(y0 , num_steps))


for step_sizes in range(20, 101 , 20):
    print(f"for Step size {step_sizes}")
    dto_grad = g_fn_DTO(y0 , step_sizes)
    otd_grad = g_fn_OTD(y0 , step_sizes)
    assert jnp.allclose(dto_grad , otd_grad)
    print(f" --> DTO memory usage: {g_fn_DTO.lower(y0 , step_sizes).compile().memory_analysis().temp_size_in_bytes}")
    print(f" --> OTD memory usage: {g_fn_OTD.lower(y0 , step_sizes).compile().memory_analysis().temp_size_in_bytes}")
    print(f"does checkpointing uses more memory than just lax scan {g_fn_DTO.lower(y0 , step_sizes).compile().memory_analysis().temp_size_in_bytes > g_fn_OTD.lower(y0 , step_sizes).compile().memory_analysis().temp_size_in_bytes} ")
```

Which means that in my case using juste a `jax.lax` is equivalent to a `BackSolveAdjoint`?

Is there anyway we can get this behaviour using the current diffrax?
I tried to comment out the custom vjp rule but I got an error 

```
Reverse-mode differentiation does not work for "
                       "lax.while_loop or lax.fori_loop with dynamic start/stop values. "
                       "Try using lax.scan, or using fori_loop with static start/stop.
```

**Comment by johannahaffner** [comment]
I'm not an expert on this, but one thing jumps out at me - without passing either a length, or a number of checkpoints, `eqxi.scan` allocates a lot more memory than required. It is in the same ballpark if you write

```python
results, _ = eqxi.scan(step, y0, jnp.arange(num_steps) , kind=kind, checkpoints=num_steps)
```
instead. Is memory-efficient integration what you are after? (I don't see any gradient computation, for which `RecursiveCheckpointAdjoint` is a better choice.)

**Comment by ASKabalan** [comment]
If I understood correctly `RecursiveCheckpointAdjoint` is a better choice for memory efficiency? 
This is not what I got from my tests

In my case I am using BackSolveAdjoint or RecursiveCheckpointAdjoint
With a some tolerance I am getting decent gradients either way (in float64)

But my question is in case I have a ConstantStepSize reversible solver
Isn't just a lax.scan more memory efficient than `RecursiveCheckpointAdjoint`

**Comment by johannahaffner** [comment]
You might like this [comment](https://github.com/patrick-kidger/diffrax/issues/564#issuecomment-2587580388) (and the thread).

**Comment by patrick-kidger** [comment]
It's unlikely that `lax.scan` is more memory efficient than `RecursiveCheckpointAdjoint`, as the latter uses only logarithmic memory and does not save intermediate values within each step. Probably you should adjust either `RecursiveCheckpointAdjoint(checkpoints=...)` or `diffeqsolve(..., max_steps=...)`. (E.g. try setting both of these to the exact number of steps you'll be taking.)

**Comment by ASKabalan** [comment]
Here are my tests

```python
import jax
import jax.numpy as jnp
from equinox.internal import scan
from typing import NamedTuple
from diffrax import diffeqsolve , Euler , ODETerm , ConstantStepSize , RecursiveCheckpointAdjoint , Euler , BacksolveAdjoint
jax.config.update("jax_enable_x64" , True)

class ODEField(NamedTuple):
    x : jnp.ndarray
    y : jnp.ndarray


def ode_funct(t , v , args):
    x = jnp.cos(v.y) * t**2
    y = jnp.sin(v.x) * t**2 
    return ODEField(x , y)

  
term = ODETerm(ode_funct)
t0 = 0.0
t1 = 5.
dt0 = 0.01

y0 = ODEField(jnp.array(1.0) , jnp.array(1.0))


@jax.jit
def diffrax_solve_dto(y0):
  solution = diffeqsolve(term , Euler() , t0=t0 , t1=t1 , dt0=dt0 , y0=y0 , stepsize_controller=ConstantStepSize() , adjoint=RecursiveCheckpointAdjoint(checkpoints=500) , args=None)

  return jax.tree.map(lambda x: x[0] , solution.ys)

@jax.jit
def diffrax_solve_otd(y0):
  solution = diffeqsolve(term , Euler() , t0=t0 , t1=t1 , dt0=dt0 , y0=y0 , stepsize_controller=ConstantStepSize() , adjoint=BacksolveAdjoint(solver=Euler()) , args=None)

  return jax.tree.map(lambda x: x[0] , solution.ys)

@jax.jit
def jax_solve(y0):
  def body_fn(v0 , t):
    dv = ode_funct(t , v0 , None)
    v1 = jax.tree.map(lambda x , dx : x + dx*dt0 , v0 , dv)
    return v1 , None

  steps = jnp.arange(t0 , t1 , dt0)
  solution , _ = scan(body_fn , y0 , steps , kind="lax")
  return solution
  
@jax.jit
def eqx_solve(y0):
  def body_fn(v0 , t):
    dv = ode_funct(t , v0 , None)
    v1 = jax.tree.map(lambda x , dx : x + dx*dt0 , v0 , dv)
    return v1 , None

  steps = jnp.arange(t0 , t1 , dt0)
  solution , _ = scan(body_fn , y0 , steps , kind="checkpointed" , checkpoints="all")
  return solution

@jax.jit
def check_jax_solve(y0):
  def body_fn(v0 , t):
    dv = ode_funct(t , v0 , None)
    v1 = jax.tree.map(lambda x , dx : x + dx*dt0 , v0 , dv)
    return v1 , None

  steps = jnp.arange(t0 , t1 , dt0)
  solution , _ = scan(jax.checkpoint(body_fn) , y0 , steps , kind="lax")
  return solution


from jax._src.ad_checkpoint import saved_residuals

def get_residual_size(f , *args):
  size = 0
  for aval , _ in saved_residuals(f , *args):
    size += aval.size * aval.dtype.itemsize
  return size
    

diffrax_dto_r = get_residual_size(diffrax_solve_dto  , y0)
diffrax_otd_r = get_residual_size(diffrax_solve_otd  , y0)
jax_r = get_residual_size(jax_solve , y0)
eqx_r = get_residual_size(eqx_solve , y0)
check_jax_r = get_residual_size(check_jax_solve , y0)
otd_jax = get_residual_size(jax_otd_solve , y0)

print(f"Residual size for diffrax dto: {diffrax_dto_r}")
print(f"Residual size for diffrax otd: {diffrax_otd_r}")
print(f"Residual size for jax: {jax_r}")
print(f"Residual size for eqx: {eqx_r}")
print(f"Residual size for check_jax: {check_jax_r}")
#Out : Residual size for diffrax dto: 961
#Out : Residual size for diffrax otd: 156
#Out : Residual size for jax: 16000
#Out : Residual size for eqx: 29048
#Out : Residual size for check_jax: 12000
```

Ok in my tests with a checkpoints = log of steps still gives decent gradients with a decent memory usage
As you said `RecursiveCheckpointAdjoint` is more memory efficient than `lax.scan` with a good choice of checkpoints
